### PR TITLE
Beam lib improvements

### PR DESCRIPTION
**Pull request by Threated**
This also adds some more rust integration tests although they don't really check any edge cases yet.
This should make it easier to port the bash tests to rust.
~~I also plan to add sockets support to the `BeamClient` in the future.~~
Future is now 🚀

### CODE REVIEW

File: beam-lib/src/http_util.rs [hunk 1]
```diff
@@ -1,21 +1,244 @@
...
+/// - `wait_count` and `wait_time` are unset => Don't wait
+/// - `wait_count` is set and `wait_time` is unset => Wait for this many items to be ready
+/// - `wait_count` is unset and `wait_time` is set => Wait for the given duration
+/// - `wait_count` is set and `wait_time` is set => Stop waiting after either of the conditions is fulfilled
+#[derive(Debug)]
+pub struct BlockingOptions {
+    /// How long to poll for
+    pub wait_time: Option<Duration>,
+    /// The number of elements to wait for
+    pub wait_count: Option<u16>,
```
Comment (TKussel) [inline beam-lib/src/http_util.rs:40]: Is there a specific reason for choosing `u16` instead of `usize`?

File: beam-lib/src/http_util.rs [hunk 2]
```diff
@@ -1,21 +1,244 @@
...
+            .await;
+        let response = match response_result {
+            Ok(res) => res,
+            Err(e) => return if e.is_timeout() {
+                Ok(Vec::with_capacity(0))
+            } else {
+                Err(e.into())
+            },
+        };
+        match response.status() {
```
Comment (TKussel) [inline beam-lib/src/http_util.rs:125]: Currently, branching based on `response.is_success()` might be clearer.

File: beam-lib/src/http_util.rs [hunk 3]
```diff
@@ -1,21 +1,244 @@
...
+    pub async fn post_task<T: Serialize>(&self, task: &TaskRequest<T>) -> Result<()> {
+        let url = self.beam_proxy_url
+            .join("/v1/tasks")
+            .expect("The proxy url is valid");
+        let response = self.client
+            .post(url)
+            .json(task)
+            .send()
+            .await?;
+        match response.status() {
```
Comment (TKussel) [inline beam-lib/src/http_util.rs:143]: As above

File: tests/src/socket_test.rs [hunk 4]
```diff
@@ -1,90 +1,35 @@
...
     const N: usize = 2_usize.pow(13);
     let test_data: &mut [u8; N] = &mut [0; N];
     rand::thread_rng().fill_bytes(test_data);
     let mut read_buf = [0; N];
-    a.write_all(test_data).await.unwrap();
-    a.flush().await.unwrap();
-    b.read_exact(&mut read_buf).await.unwrap();
+    a.write_all(test_data).await?;
+    a.flush().await?;
+    b.read_exact(&mut read_buf).await?;
```
Comment (TKussel) [inline tests/src/socket_test.rs:13]: Thanks :)

File: tests/src/tasts_test.rs
Comment (TKussel) [inline tests/src/tasts_test.rs:1]: This file should probably be named `task_test.rs`

**Review by TKussel — CHANGES_REQUESTED** [review]

File: beam-lib/src/http_util.rs (hunk 1 above)
Comment (Threated) [reply beam-lib/src/http_util.rs:40]: It is u16 in beam internally as well

File: tests/src/tasts_test.rs
Comment (Threated) [reply tests/src/tasts_test.rs:1]: Can't believe I did not catch that

File: beam-lib/src/http_util.rs (hunk 2 above)
Comment (Threated) [reply beam-lib/src/http_util.rs:125]: I disagree.
1. If we chose to add different errors in the future it will be a match anyway
2. I chose to handle a gateway timeout as returning no elements because the application logic will almost always treat these cases the same way so I need to check for the `GATEWAY_TIMEOUT` status.
3. I think it's a good idea to be really strict with what statuses we treat as valid as it will also help us catch breaking changes in beam.

File: beam-lib/src/http_util.rs (hunk 3 above)
Comment (Threated) [reply beam-lib/src/http_util.rs:143]: Here it makes a bit more sense but I still disagree. I don't think an if would really read that much nicer.

File: beam-lib/src/http_util.rs (hunk 2 above)
Comment (TKussel) [reply beam-lib/src/http_util.rs:125]: Fair points. My suggestion was based on seeing all successful status codes we process or-ed together. Your points, especially 2 and 3, are good reasons, tho.

**Review by TKussel — APPROVED** [review]
